tests/frontend/test_io.py

```diff
@@ -0,0 +1,543 @@
+        assert prog.circuit[0].op.p[0].x == 0.54
+        assert prog.circuit[0].op.p[1].x == np.pi
+        assert prog.circuit[0].reg[0].ind == 0
+        assert prog.circuit[0].reg[1].ind == 2
+
+    def test_compilation(self):
+        """Test automatic compilation using the target keyword"""
+        bb_script = """\
+        name test_program
+        version 1.0
+        target gaussian
+        Pgate(0.54) | 0
+        """
+
+        bb = blackbird.loads(bb_script)
+        prog = io.to_program(bb)
+
+        assert len(prog) == 2
+        assert prog.circuit[0].op.__class__.__name__ == "Sgate"
+        assert prog.circuit[0].reg[0].ind == 0
+        assert prog.circuit[1].op.__class__.__name__ == "Rgate"
+        assert prog.circuit[1].reg[0].ind == 0
+
+
+class TestSave:
+    """Test sf.save functionality"""
+
+    def test_save_filename_path_object(self, prog, tmpdir):
+        """Test saving a program to a file path using a path object"""
+        filename = tmpdir.join("test.xbb")
+        sf.save(filename, prog)
+
+        with open(filename, "r") as f:
+            res = f.read()
+
+        assert res == test_prog_not_compiled
+
+    def test_save_filename_string(self, prog, tmpdir):
+        """Test saving a program to a file path using a string filename"""
+        filename = str(tmpdir.join("test.xbb"))
+        sf.save(filename, prog)
+
+        with open(filename, "r") as f:
+            res = f.read()
+
+        assert res == test_prog_not_compiled
+
+    def test_save_file_object(self, prog, tmpdir):
+        """Test writing a program to a file object"""
+        filename = tmpdir.join("test.xbb")
+
+        with open(filename, "w") as f:
+            sf.save(f, prog)
+
+        with open(filename, "r") as f:
+            res = f.read()
+
+        assert res == test_prog_not_compiled
+
+    def test_save_append_extension(self, prog, tmpdir):
+        """Test appending the extension if not present"""
+        filename = str(tmpdir.join("test.txt"))
+        sf.save(filename, prog)
+
+        with open(filename+'.xbb', "r") as f:
+            res = f.read()
+
+        assert res == test_prog_not_compiled
+
+
+class TestLoad:
+    """Test sf.load functionality"""
+
+    def test_invalid_file(self, prog, tmpdir):
+        """Test exception is raised if file is invalid"""
+        with pytest.raises(ValueError, match="must be a string, path"):
+            sf.load(1)
+
+    def test_load_filename_path_object(self, prog, tmpdir):
+        """Test loading a program using a path object"""
+        filename = tmpdir.join("test.xbb")
+
+        with open(filename, "w") as f:
+            f.write(test_prog_not_compiled)
+
+        res = sf.load(filename)
+
+        # check loaded program is the same as expected
+        for cmd1, cmd2 in zip(res.circuit, prog.circuit):
+            assert cmd1.op.__class__.__name__ == cmd2.op.__class__.__name__
+            assert cmd1.reg[0].ind == cmd2.reg[0].ind
+            if cmd1.op.p:
```

Comment:
Why is the `if` necessary? Aren't these known circuits?

Reply:
Same question for two tests below

Reply:
Solely because one of the gates `Vac` has no arguments, so `cmd1.op.p == []`. This was just a quick way to exclude gates that fit this criteria (no args).